include/swift/AST/AccessScope.h

```diff
@@ -0,0 +1,82 @@
+//===--- AccessScope.h - Swift Access Scope ---------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_ACCESSSCOPE_H
+#define SWIFT_ACCESSSCOPE_H
+
+#include "swift/AST/AttrKind.h"
+#include "swift/AST/DeclContext.h"
+#include "llvm/ADT/PointerIntPair.h"
+
+namespace swift {
+
+/// The wrapper around the outermost DeclContext from which
+/// a particular declaration can be accessed.
+class AccessScope {
+  /// The declaration context (if not public) along with a bit saying
+  /// whether this scope is private (or not).
+  llvm::PointerIntPair<const DeclContext *, 1, bool> Value;
+public:
+  AccessScope(const DeclContext *DC, bool isPrivate = false)
+    : Value(DC, isPrivate) {
+      if (!DC || isa<ModuleDecl>(DC))
+        assert(!isPrivate && "public or internal scope can't be private");
+    }
+
+  static AccessScope getPublic() { return AccessScope(nullptr); }
+
+  /// Returns nullptr if access scope is public.
+  const DeclContext *getDeclContext() const { return Value.getPointer(); }
+
+  bool operator==(AccessScope RHS) const { return Value == RHS.Value; }
+  bool operator!=(AccessScope RHS) const { return !(*this == RHS); }
+  bool hasEqualDeclContextWith(AccessScope RHS) const {
+    return getDeclContext() == RHS.getDeclContext();
+  }
+
+  bool isPublic() const { return !Value.getPointer(); }
+  bool isPrivate() const { return Value.getInt(); }
+  bool isFileScope() const {
+    auto DC = getDeclContext();
+    return DC && DC->getContextKind() == DeclContextKind::FileUnit;
+  }
+
+  /// Returns true if this is a child scope of the specified other access scope.
+  ///
+  /// \see DeclContext::isChildContextOf
+  bool isChildOf(AccessScope AS) const {
+    if (!isPublic() && !AS.isPublic())
+      return getDeclContext()->isChildContextOf(AS.getDeclContext());
+    return AS.isPublic();
+  }
+
+  /// Returns the associated access level for diagnostic purposes.
+  Accessibility accessibilityForDiagnostics() const;
+
+  /// Returns the narrowest access scope if this and the specified access scope
+  /// have common intersection, or None if scopes don't intersect.
+  const Optional<AccessScope> intersectWith(AccessScope accessScope) const {
+    if (hasEqualDeclContextWith(accessScope))
+      return *this;
```

Comment:
What do you think we should do if we're intersecting `fileprivate` with `private`-at-file-scope? I feel like we should pick one or the other, rather than just stick with what's on the left-hand side, but which one do you think makes more sense to show up in diagnostics?